fix(ivy): ngcc - several dependency related fixes

JoostK opened the pull request:
**fix(ivy): ngcc - prevent crash when analyzed target is ignored**

ngcc analyzes the dependency structure of the entrypoints it needs to
process, as the compilation of entrypoints is ordering sensitive: any
dependent upon entrypoint must be compiled before its dependees. As part
of the analysis of the dependency graph, it is detected when a
dependency of entrypoint is not installed, in which case that entrypoint
will be marked as ignored.

When a target entrypoint to compile is provided, it could occur that
given target is considered ignored because one of its dependencies might
be missing. This situation was not dealt with currently, instead
resulting in a crash of ngcc.

This commit prevents the crash by taking the above scenario into account.

---

**fix(ivy): ngcc - do not consider builtin NodeJS modules as missing**

ngcc analyzes the dependency structure of the entrypoints it needs to
process, as the compilation of entrypoints is ordering sensitive: any
dependent upon entrypoint must be compiled before its dependees. As part
of the analysis of the dependency graph, it is detected when a
dependency of entrypoint is not installed, in which case that entrypoint
will be marked as ignored.

For libraries that work with Angular Universal to run in NodeJS, imports
into builtin NodeJS modules can be present. ngcc's dependency analyzer
can only resolve imports within the TypeScript compilation, which
builtin modules are not part of. Therefore, such imports would
erroneously cause the entrypoint be become ignored.

This commit fixes the problem by taking the NodeJS builtins into account
when dealing with missing imports.

Fixes #31522

JoostK committed:
fix(ivy): ngcc - prevent crash when analyzed target is ignored

ngcc analyzes the dependency structure of the entrypoints it needs to
process, as the compilation of entrypoints is ordering sensitive: any
dependent upon entrypoint must be compiled before its dependees. As part
of the analysis of the dependency graph, it is detected when a
dependency of entrypoint is not installed, in which case that entrypoint
will be marked as ignored.

When a target entrypoint to compile is provided, it could occur that
given target is considered ignored because one of its dependencies might
be missing. This situation was not dealt with currently, instead
resulting in a crash of ngcc.

This commit prevents the crash by taking the above scenario into account.

JoostK committed:
fix(ivy): ngcc - do not consider builtin NodeJS modules as missing

ngcc analyzes the dependency structure of the entrypoints it needs to
process, as the compilation of entrypoints is ordering sensitive: any
dependent upon entrypoint must be compiled before its dependees. As part
of the analysis of the dependency graph, it is detected when a
dependency of entrypoint is not installed, in which case that entrypoint
will be marked as ignored.

For libraries that work with Angular Universal to run in NodeJS, imports
into builtin NodeJS modules can be present. ngcc's dependency analyzer
can only resolve imports within the TypeScript compilation, which
builtin modules are not part of. Therefore, such imports would
erroneously cause the entrypoint to become ignored.

This commit fixes the problem by taking the NodeJS builtins into account
when dealing with missing imports.

Fixes #31522

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts:
```diff
@@ -81,7 +81,7 @@ export class DependencyResolver {
     let sortedEntryPointNodes: string[];
     if (target) {
-      if (target.compiledByAngular) {
+      if (target.compiledByAngular && graph.hasNode(target.path)) {
```
If we have specifically requested a target entry-point then this **should** be a fatal error, not silently ignored.
Currently we crash, instead we should log a useful error explaining which dependency is missing.

petebacondarwin commented on packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts:
```diff
@@ -128,10 +128,12 @@ export class DependencyResolver {
       const {dependencies, missing, deepImports} = this.getEntryPointDependencies(entryPoint);
 
-      if (missing.size > 0) {
+      const missingDependencies = computeMissingDependencies(missing);
```
I find `computeMissingDependencies()` as a function name confusing. We have already computed the missing ones. What this function is doing is filtering out ignorable ones.

```ts
const missingDependencies = filterIgnorableDependencies(missing);
```

Alternatively we could not even call a function here but simply to inline the filtering. E.g.

```ts
const missingDependencies = Array.from(missing).filter(dep => !ignoredImports.has(dep));
```

JoostK replied:
Changed into the inline variant, it's one-liner implementation is readable enough :-)

petebacondarwin reviewed (changes requested):
Thanks @JoostK.

A nit in the second commit, but I think the first commit should throw a helpful fatal error.

gkalpak commented on packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts:
```diff
@@ -191,3 +193,14 @@ export class DependencyResolver {
   graph: DepGraph<EntryPoint>;
 }
+
+const ignoredImports = new Set<string>(require('module').builtinModules);
```
I prefer to have such constants at the top of the file (before any class definitions), because they are easy to miss down here. (Also, `ignoredImports` is little anbiguous. I would consider renaming to something like `ignorableImports` or `builtInNodeJsModules` or something.)

JoostK replied:
Moved it to the top, and renamed.

gkalpak commented on packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts:
```diff
@@ -191,3 +193,14 @@ export class DependencyResolver {
   graph: DepGraph<EntryPoint>;
 }
+
+const ignoredImports = new Set<string>(require('module').builtinModules);
```
OOC, in there a reason for using `require` instead of `import`?

JoostK replied:
I got an error when using `import {builtinModules} from 'module';`:

```
Error:(15, 22) TS2497: This module can only be referenced with ECMAScript imports/exports by turning on the 'esModuleInterop' flag and referencing its default export.
```

The declaration of the `module` module is special:

```
declare module "module" {
    export = NodeJS.Module;
}
```

gkalpak reviewed (commented):
👌

Typo in the 2nd commit message: `be become ignored` --> `to become ignored`

petebacondarwin committed:
fixup! fix(ivy): ngcc - prevent crash when analyzed target is ignored

petebacondarwin committed:
fix(ivy): ngcc - report an error if a target has missing dependencies

Previously, we either crashed with an obscure error or silently did
nothing. Now we throw an exception but with a helpful message.

JoostK committed:
fixup! fix(ivy): ngcc - do not consider builtin NodeJS modules as missing

petebacondarwin commented:
merge-assistance: the commits are all CLA checked
